WIP sql: use singleflight.Group to perform table name resolution

vivekmenezes opened the pull request:
Use singleflight.Group to perform table name resolution

fixes #36044

Release note: None

vivekmenezes committed:
sql: refactor AcquireByName to clarify when a name is being resolved

This simplifies the code by getting rid of the call to
AcquireFreshestFromStore and splitting name resolution into
two calls: one done at the current time, and the other at the
requested timestamp.

In a future change name resolution done at the current time
will be done once for all queries using this API through the
use of a singleflight.Group

This changes is tested by the tests in rename_test.go

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/36243)
<!-- Reviewable:end -->

vivekmenezes committed:
sql: resolve table name -> id only in a single thread

this change will block all other threads wanting to resolve the
table name -> id mapping thereby resolving the name only by a single
node. The system.namespace table will see a single request from
a node.

Release note: None

vivekmenezes commented:
will un-WIP this once I know this fixes the issue
